### ISSUE TITLE
More mature / graceful ABI support

### ISSUE THREAD

**Issue by tkporter** [FuelLabs/fuel-indexer#453]
I'm unable to use my contract's ABIs with the indexer - the abi isn't too exotic and works with the existing fuels-rs abigen and fuels-ts tooling.

It'd be great to have more types supported by the indexer macro, but I understand this can be a pretty large undertaking. It would be really nice to be able to at least easily opt out of certain types and to still use my contract's ABI for the types that are supported. My only options are to either forego using an ABI entirely, or to manually strip out the problematic types from my ABI, which is super fragile.

This is a non-exhaustive list of the types I'm having issues with:

A tuple:

From my ABI:
```
    {
      "typeId": 1,
      "type": "(_, _)",
      "components": [
        {
          "name": "__tuple_element",
          "type": 2,
          "typeArguments": null
        },
        {
          "name": "__tuple_element",
          "type": 21,
          "typeArguments": null
        }
      ],
      "typeParameters": null
    },
```
And the error:
```
error: custom attribute panicked
  --> indexer/mailbox/src/lib.rs:58:1
   |
58 | #[indexer(manifest = "fuel-contracts/indexer/mailbox/mailbox.manifest.yaml")]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: message: `"_)"` is not a valid identifier
```

`generic T` resulting from usage of Option in my contract's interface. From my ABI:
```
    {
      "typeId": 5,
      "type": "enum Option",
      "components": [
        {
          "name": "None",
          "type": 0,
          "typeArguments": null
        },
        {
          "name": "Some",
          "type": 6,
          "typeArguments": null
        }
      ],
      "typeParameters": [
        6
      ]
    },
    {
      "typeId": 6,
      "type": "generic T",
      "components": null,
      "typeParameters": null
    },
```

The error:
```
error: Unrecognized primitive type: "generic T".
  --> indexer/mailbox/src/lib.rs:58:1
   |
58 | #[indexer(manifest = "fuel-contracts/indexer/mailbox/mailbox.manifest.yaml")]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `indexer` (in Nightly builds, run with -Z macro-backtrace for more info)
```

`raw untyped ptr` due to `RawVec` which is due to `Vec`. The error:
```
error: Unrecognized primitive type: "raw untyped ptr".
  --> indexer/mailbox/src/lib.rs:58:1
   |
58 | #[indexer(manifest = "fuel-contracts/indexer/mailbox/mailbox.manifest.yaml")]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `indexer` (in Nightly builds, run with -Z macro-backtrace for more info)
```

We also plan to use the new-ish `Bytes` type in favor of `Vec` once https://github.com/FuelLabs/fuels-rs/issues/723 is completed, which would also use `raw untyped ptr` under the hood

**Comment by ra0x3** [comment]
Hey thanks for this @tkporter. 
- Obviously the indexer supports the more fundamental ABI types, but I see you're using some of the newer stuff here (which we obviously don't support, yet). 
- I definitely understand it can be frustrating to write a valid Sway contract, but then those contract types not work with the indexer
- You sort've hit the nail on the head with "..but I understand this can be a pretty large undertaking..". 
  - The Sway team is huge so they're cranking out increasingly new/more features 👍, and it's hard for the indexer team (a team of 2) to keep up with the output 👎 
- The quick/immediate fix is just to clarify what Sway functionality we do/don't support -- and then address some of this internally

**Comment by tkporter** [comment]
thanks for the reply, I totally understand that there's a lot of flux and a lot of moving pieces :)
Also didn't mean to bombard you guys with a bunch of issues lol - I spent the last few days getting our indexer up and running and in the process wanted to flag the bits that I found difficult. Things are working quite well without our specific ABI so this is definitely a nice-to-have for us

**Comment by ra0x3** [comment]
Blocked by https://github.com/FuelLabs/fuel-indexer/issues/423

**Comment by ra0x3** [comment]
Hey @tkporter j
- Just wanted to follow up and hopefully provide a little bit of closure/clarity here.
- In addition to upgrading our dependencies and getting onto the latest Sway version (this issue https://github.com/FuelLabs/fuel-indexer/issues/423), I saw that this particular issue is related specifically to the `tuple` type in Sway
- I've opened a spike issue https://github.com/FuelLabs/fuel-indexer/issues/504 in order to sus out what support for specifically tuples might look like
- So far unfortunately, I'm seeing that [GraphQL seems to not support tuple types out of the box](https://github.com/graphql/graphql-spec/issues/534)- from what I can see
  - Further, since we build our database schema using GraphQL schema - if GraphQL doesn't support it, we wouldn't be able to support it
    - However, this only applies to data going into the database -- after https://github.com/FuelLabs/fuel-indexer/issues/504 you should at least be able to still use `tuple`s from your ABI JSON in your index (just not persist them to the database)
  
So with all that said @tkporter were you planning to save the tuple type to the database? Or were you just trying to use the value in your index (without saving to the DB)? 

Hope this helps

**Comment by tkporter** [comment]
Hey, appreciate the info! We were just hoping to use the value in our index without saving to the DB. The main goal of ours was to be able to pass in the ABI without compilation breaking, and from there we only really need access to the more "vanilla" types

**Comment by ra0x3** [comment]
@tkporter Thanks for getting back to me.
- So `tuple` support was merged with https://github.com/FuelLabs/fuel-indexer/pull/506
- We should be cutting a v0.2 release here shortly at which time, you should be able to use this functionality
- As mentioned, this `tuple` support won't work with the data layer (and GraphQL schema) but you should definitely be able to access your tuple items from your index handler function
- Just for a quick reference [here](https://github.com/FuelLabs/fuel-indexer/blob/master/packages/fuel-indexer-tests/contracts/fuel-indexer-test/src/main.sw#L137) is how we support tuples coming from Sway smart contracts, and [here](https://github.com/FuelLabs/fuel-indexer/blob/master/packages/fuel-indexer-tests/components/indices/fuel-indexer-test/src/lib.rs#L236) is how you can pull that tuple into your index handler function

Let me know if you have any questions - happy to open this back up as well (it was closed with the associated PR was merged)
